### ISSUE TITLE
Support memoizing via instance variables without pre-declaring their types

### ISSUE THREAD

**Issue by viraptor** [sorbet/sorbet#995]
### Problem

In strict mode, instance variables need to be declared early. This is quite verbose in a common case of a caching variable in a method. For example:

```
sig {returns(Foo)}
def foo
  @foo ||= generate_foo
end
```

is using a very common idiom. However it gets error:

```
Use of undeclared variable @foo
```

### Proposed solution

Potentially special case methods in form:
```
def <name>
  @<name> ||= ...
end
```
so they don't need the extra variable declaration. This would be useful, because when using this idiom, `@<name>` is normally never used outside of that single statement.

**Comment by DarkDimius** [comment]
Note that `strict` level is expected to be more verbose than `typed: true`. In this level you're effectively saying that you are volunteering for more types and are volunteering to declare _all_ of your methods and fields.

**Comment by viraptor** [comment]
Yes, I understand that's the expected behaviour at the moment. I think the specific pattern could be special-cased since it's really common. But I understand you may prefer consistency.

**Comment by jez** [comment]
For what it's worth, I don't find it **that** intrusive having to pre-declare the instance variable's type. For example, given this:

```ruby
# typed: strict

module Util
  extend T::Sig

  sig { returns(String) }
  def self.git_sha
    @git_sha ||= ENV.fetch(
      'GIT_REV',
      `git rev-parse --verify HEAD 2>&1`
    ).chomp
  end
end
```

The only change needed is to add a type for `@git_sha` above the method:

```ruby
  # ...
  @git_sha = T.let(nil, T.nilable(String))
  sig {...}
```

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20strict%0A%23%20frozen_string_literal%3A%20true%0A%0Amodule%20Util%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20%40git_sha%20%3D%20T.let(nil%2C%20T.nilable(String))%0A%0A%20%20sig%20%7B%20returns(String)%20%7D%0A%20%20def%20self.git_sha%0A%20%20%20%20%40git_sha%20%7C%7C%3D%20ENV.fetch(%0A%20%20%20%20%20%20%27GIT_REV%27%2C%0A%20%20%20%20%20%20%60git%20rev-parse%20--verify%20HEAD%202%3E%261%60%0A%20%20%20%20).chomp%0A%20%20end%0Aend)

which is akin to adding a sig onto a method, and has the added benefit that if you use `@git_sha` anywhere else in the file, Sorbet will catch you if you forget to check that it might not be initialized.

**Comment by jez** [comment]
^ for this reason I think we're going to close this for the time being, because we're not **that** convinced it's painful right now, and because any implementation change to support this would be invasive.

**Comment by hudon** [comment]
^ The above only works for class methods, though.

Below is a real-world example of the pain this issue causes. The whole `initialize` method has to be defined and we no longer have the advantage subclassing brings that is to not have to know how the parent class is instantiated.

```
class Model
  sig { params(foo: String, bar: String, baz: Integer).void }
  def initialize(foo, bar, baz: 5)
  end
end

class MyModel < Model
  # This whole initialize definition cannot be avoided
  sig { params(foo: String, bar: String, baz: Integer).void }
  def initialize(foo, bar, baz: 5)
    super(foo, bar, baz: baz)
    @api = T.let(nil, T.nilable(Api))
  end

  sig { void }
  def do_foo
    api.call_foo
  end

  private
  
  sig { returns(Api) }
  def api
    @api ||= Api.new
  end
end

class Api
  sig { void }
  def call_foo
  end
end
```

I tried various approaches found in the docs to get around this, but I may have missed something. Is there a better way?

**Comment by hudon** [comment]
actually, we can probably use `before_initialize` or `after_initialize` to avoid the above issue. Nevermind!

**Comment by RobinDaugherty** [comment]
The example given to declare the property only works in a module, not for instance methods on a class.

Having to pre-initialize these properties is incredibly awkward and decouples the declaration from the code that uses the property.

The property type should be explictly known to Sorbet assuming there's a method return signature. Wrapping that in nilability is something that Sorbet should do intuitively, especially considering memoization like this is a common Ruby practice.

This would even be an opportunity for Sorbet to detect and warn when a nilable return value is declared on a memoized method value, since that might cause the memoized operation to be called multiple times.

On the other hand, I do understand this is only a "problem" when strict log level is enabled, but that doesn't seem like a satisfying answer considering that we want to encourage strict log level since it's the style of use that provides the most value for use of Sorbet. Coupled with the fact that memoization is a common practice, it seems shortsighted for Sorbet not to provide a good solution for it.

**Comment by amomchilov** [comment]
For the record, support for this was implemented in #6081, and released in [`0.5.10210.20220725105033-6e80c502e`](https://github.com/sorbet/sorbet/releases/tag/0.5.10210.20220725105033-6e80c502e)

🥳
